netket/utils/group/group.py

```diff
@@ -0,0 +1,321 @@
+    two Elements are considered equal iff the corresponding matrices are equal.
+
+    """
+
+    def __hash__(self):
+        return super().__hash__()
+
+    def __matmul__(self, other) -> "Group":
+        if not isinstance(other, Group):
+            raise ValueError("`Group`s can only be multiplied with other `Group`s")
+        return Group(super().__matmul__(other).elems)
+
+    def _canonical(self, x: Element) -> Array:
+        """
+        Canonical form of :code:`Element`s, used for equality testing (i.e., two :code:`Element`s
+        `x,y` are deemed equal iff :code:`_canonical(x) == _canonical(y)`.
+        Must be overridden in subclasses
+
+        Arguments:
+            x: an `Element`
+
+        Returns:
+            the canonical form as a numpy.ndarray of integers
+        """
+        raise NotImplementedError
+
+    def _canonical_array(self) -> Array:
+        """
+        Lists the canonical forms returned by `_canonical` as rows of a 2D array.
+        """
+        return np.array([self._canonical(x).flatten() for x in self.elems])
+
+    def _canonical_lookup(self) -> dict:
+        """
+        Creates a lookup table from canonical forms to index in `self.elems`
+        """
+        return {
+            HashableArray(self._canonical(element)): index
+            for index, element in enumerate(self.elems)
+        }
+
+    def remove_duplicates(self, *, return_inverse=False) -> "Group":
+        """
+        Returns a new :code:`Group` with duplicate elements (that is, elements with
+        identical canonical forms) removed.
+
+        Arguments:
+            return_inverse: If True, also return indices to reconstruct the original
+                group from the result.
+
+        Returns:
+            group: the group with duplicate elements removed.
+            inverse: Indices to reconstruct the original group from the result.
+                Only returned if `return_inverse` is True.
+        """
+        result = np.unique(
+            self._canonical_array(),
+            axis=0,
+            return_index=True,
+            return_inverse=return_inverse,
+        )
+        group = Group([self.elems[i] for i in sorted(result[1])])
+        if return_inverse:
+            return group, result[2]
+        else:
+            return group
+
+    @struct.property_cached
+    def inverse(self) -> Array:
+        """
+        Indices of the inverse of each element.
+        If :code:`g = self[idx_g]` and :code:`h = self[self.inverse[idx_g]]`, then
+        :code:`gh = product(g, h)` is equivalent to :code:`Identity()`
+        """
+        canonical_identity = self._canonical(Identity())
+        inverse = np.zeros(len(self.elems), dtype=int)
+
+        for i, e1 in enumerate(elems):
+            for j, e2 in enumerate(elems):
+                prod = e1 @ e2
+                if np.all(self._canonical(prod) == canonical_identity):
+                    inverse[i] = j
```

Comment:
This cannot work because it should be `self.elems`. The fact that the tests pass regardless means we need to add a test for `inverse`, at least.

```suggestion
        for i, e1 in enumerate(self.elems):
            for j, e2 in enumerate(self.elems):
                prod = e1 @ e2
                if np.all(self._canonical(prod) == canonical_identity):
                    inverse[i] = j
```

Reply:
Oops, thanks! It's kinda hard to test `Group.inverse` and `product_table` directly since every exposed group overrides them...